library/pk.c

```diff
@@ -232,6 +232,126 @@ int mbedtls_pk_can_do( const mbedtls_pk_context *ctx, mbedtls_pk_type_t type )
     return( ctx->pk_info->can_do( type ) );
 }
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+/*
+ * Tell if a PK can do the operations of the given PSA algorithm
+ */
+int mbedtls_pk_can_do_ext( const mbedtls_pk_context *ctx, psa_algorithm_t alg,
+                           psa_key_usage_t usage )
+{
+    psa_key_usage_t key_usage;
+
+    /* A context with null pk_info is not set up yet and can't do anything.
+     * For backward compatibility, also accept NULL instead of a context
+     * pointer. */
+    if( ctx == NULL || ctx->pk_info == NULL )
+        return( 0 );
+
+    /* Filter out non allowed algorithms */
+    if( PSA_ALG_IS_ECDSA( alg ) == 0 &&
+        PSA_ALG_IS_RSA_PKCS1V15_SIGN( alg ) == 0 &&
+        PSA_ALG_IS_RSA_PSS( alg ) == 0 &&
+        alg != PSA_ALG_RSA_PKCS1V15_CRYPT &&
+        PSA_ALG_IS_ECDH( alg ) == 0 )
+        return( 0 );
```

Comment:
Should also document this in pk.h, e.g. "return 0 for non-allowed algorithms and usage flags"

Reply:
We may want to expand the set of allowed algorithms & usage flags in the future, so applications should not rely on the fact the `pk_can_do_ext( ..., PSA_KEY_USAGE_ENCRYPT )` always returns 0, because it might start returning 1 at some point in the future (if and when we implement public operations with opaque keys).

So, either we document that the return value for non-allowed algs & flags is undefined (which may sound unfriendly, UB is not popular) or we add a prominent warning that the set of allowed algs & flags may be expanded in the future, so you must not rely on the return value being 0 for algs & flags that are not supported today, because that may change in the future.